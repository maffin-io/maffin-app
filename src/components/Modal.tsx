import React from 'react';

export type ModalProps = {
  open: boolean;
  setOpen: Function;
  title?: string;
  children: React.ReactNode;
};

export default function Modal({
  open, setOpen, title, children,
}: ModalProps): JSX.Element {
  if (!open) {
    return <span />;
  }

  return (
    <div
      className="fixed inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50"
      id="my-modal"
    >
      <div
        className="relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700"
      >
        <button
          type="button"
          className="float-right"
          onClick={() => setOpen(false)}
        >
          X
        </button>
        {
          title
          && (
            <div className="w-full pb-4 border-b border-gunmetal-600">
              <span className="text-lg">
                {title}
              </span>
            </div>
          )
        }
        {children}
      </div>
    </div>
  );
}
