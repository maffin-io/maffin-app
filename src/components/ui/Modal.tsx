import React from 'react';
import RModal from 'react-modal';

export type ModalProps = {
  className?: string;
  triggerContent: string | JSX.Element;
  triggerClassName?: string;
  showClose?: boolean;
} & React.PropsWithChildren;

export interface ModalRef {
  openModal: () => void;
  closeModal: () => void;
}

function Modal({
  className = 'modal',
  triggerClassName = '',
  triggerContent,
  showClose = false,
  children,
}: ModalProps, ref: React.Ref<ModalRef>): JSX.Element {
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  // Expose actions to the parent
  React.useImperativeHandle(ref, () => ({
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  }));

  return (
    <>
      <RModal
        isOpen={isOpen}
        overlayClassName="overlay"
        className={className}
        onRequestClose={() => setIsOpen(false)}
      >
        {
          showClose
          && (
            <button
              type="button"
              className="float-right"
              onClick={() => setIsOpen(false)}
            >
              X
            </button>
          )
        }
        {children}
      </RModal>

      <button
        type="button"
        onClick={() => setIsOpen(!isOpen)}
        className={triggerClassName}
      >
        {triggerContent}
      </button>
    </>
  );
}

export default React.forwardRef(Modal);
