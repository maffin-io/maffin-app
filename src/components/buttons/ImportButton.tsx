import React from 'react';
import Modal from 'react-modal';
import { BiImport } from 'react-icons/bi';

import DBImportButton from './import/DBImportButton';
import PlaidImportButton from './import/PlaidImportButton';

export interface ImportButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  className?: string;
  onImport?: Function;
}

export default function ImportButton({
  className = 'btn btn-primary',
  onImport,
  ...props
}: ImportButtonProps): JSX.Element {
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);

  return (
    <>
      {/* @ts-ignore */}
      <Modal
        isOpen={isModalOpen}
        overlayClassName="overlay"
        className="modal bg-background-800"
      >
        <button
          type="button"
          className="float-right"
          onClick={() => setIsModalOpen(false)}
        >
          X
        </button>
        <div className="grid grid-cols-2 justify-center mt-5">
          <DBImportButton onImport={() => setIsModalOpen(false)} />
          <PlaidImportButton onImport={() => setIsModalOpen(false)} />
        </div>
      </Modal>
      <button
        id="import-button"
        type="button"
        className={className}
        onClick={() => setIsModalOpen(!isModalOpen)}
        {...props}
      >
        <BiImport className="inline-block align-middle mr-1" />
        <span className="inline-block align-middle">Import</span>
      </button>
    </>
  );
}
