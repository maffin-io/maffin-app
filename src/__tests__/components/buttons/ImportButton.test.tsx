import React from 'react';
import {
  render,
  screen,
  fireEvent,
} from '@testing-library/react';
import Modal from 'react-modal';

import ImportButton from '@/components/buttons/ImportButton';

jest.mock('react-modal', () => jest.fn(
  (props: React.PropsWithChildren) => (
    <div data-testid="Modal">
      {props.children}
    </div>
  ),
));

jest.mock('@/components/buttons/import/DBImportButton', () => jest.fn(
  () => <div data-testid="DBImportButton" />,
));

describe('ImportButton', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders hidden modal on mount', async () => {
    const { container } = render(<ImportButton />);

    expect(Modal).toBeCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );

    expect(container).toMatchSnapshot();
  });

  it('opens modal when clicking the button', async () => {
    render(<ImportButton />);

    const button = await screen.findByRole('button', { name: 'Import' });
    fireEvent.click(button);

    const modal = await screen.findByTestId('Modal');
    expect(Modal).toBeCalledWith(
      expect.objectContaining({
        isOpen: true,
      }),
      {},
    );
    expect(modal).toMatchSnapshot();
  });

  it('closes modal when clicking the X button', async () => {
    render(<ImportButton />);

    const button = await screen.findByRole('button', { name: 'Import' });
    fireEvent.click(button);

    await screen.findByTestId('Modal');

    const xButton = screen.getByRole('button', { name: 'X' });
    fireEvent.click(xButton);

    expect(Modal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );
  });
});
