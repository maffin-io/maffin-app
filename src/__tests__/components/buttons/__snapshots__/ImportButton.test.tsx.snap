// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportButton opens modal when clicking the button 1`] = `
<div
  data-testid="Modal"
>
  <button
    class="float-right"
    type="button"
  >
    X
  </button>
  <div>
    <div
      data-testid="DBImportButton"
    />
  </div>
</div>
`;

exports[`ImportButton renders hidden modal on mount 1`] = `
<div>
  <div
    data-testid="Modal"
  >
    <button
      class="float-right"
      type="button"
    >
      X
    </button>
    <div>
      <div
        data-testid="DBImportButton"
      />
    </div>
  </div>
  <button
    class="btn btn-primary"
    id="import-button"
    type="button"
  >
    <svg
      class="inline-block align-middle mr-1"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m12 18 4-5h-3V2h-2v11H8z"
      />
      <path
        d="M19 9h-4v2h4v9H5v-9h4V9H5c-1.103 0-2 .897-2 2v9c0 1.103.897 2 2 2h14c1.103 0 2-.897 2-2v-9c0-1.103-.897-2-2-2z"
      />
    </svg>
    <span
      class="inline-block align-middle"
    >
      Import
    </span>
  </button>
</div>
`;
